File: solutions/javascript/dfs/numberOfConnectedComponentsInAnUndirectedGraph.js
const buildGraph = (edges) => {
    const graph = {};

    for (const edge of edges) {
        const [nodeOne, nodeTwo] = edge;

        const nodeOneInGraph = graph.hasOwnProperty(nodeOne);
        const nodeTwoInGraph = graph.hasOwnProperty(nodeTwo);

        if (!nodeOneInGraph) graph[nodeOne] = [];
        if (!nodeTwoInGraph) graph[nodeTwo] = [];

        graph[nodeOne].push(nodeTwo);
        graph[nodeTwo].push(nodeOne);
    }

    return graph;
};

const markComponentAsVisited = (graph, node, visited) => {
    // Base cases
    if (visited.has(node)) return;

    // Process node
    visited.add(node);

    // Add neighbors
    const nodeInGraph = graph.hasOwnProperty(node);
    if (!nodeInGraph) return;

    const neighbors = graph[node];

    for (const neighbor of neighbors) {
        markComponentAsVisited(graph, neighbor, visited);
    }
};

const countComponents = (numNodes, edges) => {
    const graph = buildGraph(edges);

    let numConnectedComponents = 0;
    const visited = new Set();

    for (let node = 0; node < numNodes; node++) {
        if (visited.has(node)) continue;

        numConnectedComponents++;
        markComponentAsVisited(graph, node, visited);
    }

    return numConnectedComponents;
};

File: solutions/python/dfs/number_of_connected_components_in_an_undirected_graph.py
from collections import defaultdict

def build_graph(edges):
    graph = defaultdict(list)

    for edge in edges:
        node_one, node_two = edge

        graph[node_one].append(node_two)
        graph[node_two].append(node_one)

    return graph

class Solution:
    def mark_component_as_visited(self, graph, node, visited):
        # Base cases
        if node in visited:
            return

        # Process node
        visited.add(node)

        # Add neighbors
        if node not in graph:
            return

        neighbors = graph[node]

        for neighbor in neighbors:
            self.mark_component_as_visited(graph, neighbor, visited)

    def count_components(self, num_nodes, edges):
        graph = build_graph(edges)

        num_connected_components = 0
        visited = set()

        for node in range(num_nodes):
            if node in visited:
                continue

            num_connected_components += 1
            self.mark_component_as_visited(graph, node, visited)

        return num_connected_components

    def countComponents(self, num_nodes: int, edges: List[List[int]]) -> int:
        graph = build_graph(edges)

        num_connected_components = 0
        visited = set()

        for node in range(num_nodes):
            if node in visited:
                continue

            num_connected_components += 1
            self.mark_component_as_visited(graph, node, visited)

        return num_connected_components